addons/website_sale_stock/controllers/main.py

```diff
@@ -0,0 +1,32 @@
+# -*- coding: utf-8 -*-
+# Part of Odoo. See LICENSE file for full copyright and licensing details.
+import json
+import logging
+from datetime import datetime
+from werkzeug.exceptions import Forbidden, NotFound
+from odoo.addons.website_sale.controllers.main import WebsiteSale
+
+from odoo import fields, http, SUPERUSER_ID, tools, _
+from odoo.http import request
+from odoo.exceptions import ValidationError,UserError
+from odoo.osv import expression
+
+_logger = logging.getLogger(__name__)
+
+
+class WebsiteSaleStock(WebsiteSale):
+    @http.route(['/shop/payment/transaction/',
+        '/shop/payment/transaction/<int:so_id>',
+        '/shop/payment/transaction/<int:so_id>/<string:access_token>'], type='json', auth="public", website=True)
+    def payment_transaction(self, **kwargs):
+        """ Payment transaction override to double check cart quantities before
+        placing the order
+        """
+        order = request.website.sale_get_order()
+        values = {}
+        for line in order.order_line:
+            if line.exists():
+                values.update(order._cart_update(product_id=line.product_id.id, line_id=line.id, add_qty=0))
+        if values['warning']:
+            raise ValidationError("Some products became unavailable and your cart has been updated. We're sorry for the inconvenience.")
```

Comment:
@nle-odoo 
I'm trying to get this override of payment controller working, but I cannot find a working way of giving back an error message without making the view stuck...
Basically this is the first step executed when someone click on 'Pay Now' button, so I was trying to intercept the call and make a final check before the payment occur.
May you have some idea on how to redirect the user to the cart with a warning?

Reply:
I'm not sure, usually we would get the error before clicking on "Pay now" 

 if a product becomes unavailable, shouldn't there be a warning:

https://github.com/odoo/odoo/blob/90008eaee70744a6080ea2e7359a0bf0a6eb7714/addons/website_sale_stock/models/sale_order.py#L27-L30

on the cart already:

https://github.com/odoo/odoo/blob/90008eaee70744a6080ea2e7359a0bf0a6eb7714/addons/website_sale_stock/views/website_sale_stock_templates.xml#L9-L12

if yes you could just redirect (`return request.redirect(...)`)

Reply:
```
        if values.get('warning'):
            return request.redirect('/shop/cart')
```
will just stuck the view :/